applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py

```diff
@@ -49,15 +132,154 @@ def FinalizeAfterOptimizationLoop(self):
+        for response_id, dependencies, _ in dependencies:
+            if len(dependencies) > 0:
+                self.__RequestGradientsRecursively(dependencies, communicator)
+            else:
+                communicator.requestGradientOf(response_id)
+
+    # --------------------------------------------------------------------------
+    def __CombineResponsesAccordingDependencies(self, communicator):
+        for response_id, dependencies, _ in self.dependency_graph:
+            if len(dependencies) > 0:
+                if communicator.isRequestingValueOf(response_id):
+                    combined_value = self.__ComputeCombinedValuesRecursively(dependencies, communicator)
+                    communicator.reportValue(response_id, combined_value)
+
+        for response_id, dependencies, _ in self.dependency_graph:
+            if len(dependencies) > 0:
+                if communicator.isRequestingGradientOf(response_id):
+                    combined_gradient = self.__ComputeCombinedGradientsRecursively(dependencies, communicator)
+                    communicator.reportGradient(response_id, combined_gradient)
+
+    # --------------------------------------------------------------------------
+    def __ComputeCombinedValuesRecursively(self, dependencies, communicator):
+        combined_value = 0.0
+
+        for response_id, dependencies, weight in dependencies:
+            if len(dependencies) > 0:
+                value = self.__ComputeCombinedValuesRecursively(dependencies, communicator)
+                communicator.reportValue(response_id, value)
+            else:
+                value = communicator.getStandardizedValue(response_id)
+            combined_value += weight*value
+
+        return combined_value
+
+    # --------------------------------------------------------------------------
+    def __ComputeCombinedGradientsRecursively(self, dependencies, communicator):
+        combined_gradient = None
+
+        for response_id, dependencies, weight in dependencies:
+            if len(dependencies) > 0:
+                gradient = self.__ComputeCombinedGradientsRecursively(dependencies, communicator)
+                communicator.reportGradient(response_id, gradient)
+            else:
+                gradient = communicator.getStandardizedGradient(response_id)
+
+            gradient = {key: [weight*value[0],weight*value[1],weight*value[2]] for key, value in gradient.items()}
+            if combined_gradient is None:
+                combined_gradient = gradient
+            else:
+                # Perform nodal sum
+                combined_gradient = {key_a: [a+b for a, b in zip(list_a, list_b)] for ((key_a, list_a),(key_b, list_b)) in zip(combined_gradient.items(), gradient.items())}
+
+        return combined_gradient
+
+    # --------------------------------------------------------------------------
+    def __ComputeGradientNormsRecursively(self, dependencies, communicator):
```

Comment:
This and the following functions simply work on all the existing responses. this means they do not have to be ```Recursively``` but could simply accept a list of response identifiers and then do their work on each of those responses individually by getting what they need from the communicator.